src/dict.c

```diff
@@ -742,12 +742,24 @@ dictEntry *dictUnlink(dict *d, const void *key) {
     return dictGenericDelete(d, key, 1);
 }
 
+inline static void dictFreeKey(dictType *type, dictEntry *entry) {
+    if (type->keyDestructor) {
+        type->keyDestructor(dictGetKey(entry));
+    }
+}
+
+inline static void dictFreeVal(dictType *type, dictEntry *entry) {
+    if (type->valDestructor) {
+        type->valDestructor(dictGetVal(entry));
+    }
+}
+
 /* You need to call this function to really free the entry after a call
  * to dictUnlink(). It's safe to call this function with 'he' = NULL. */
 void dictFreeUnlinkedEntry(dict *d, dictEntry *he) {
     if (he == NULL) return;
-    dictFreeKey(d, he);
-    dictFreeVal(d, he);
+    dictFreeKey(d->type, he);
+    dictFreeVal(d->type, he);
```

Comment:
These functions are called in three places. Taking `d->type` in all three places, before calling the function, looks like code duplication to me. It's better to pass `d` and let the receiving function do it.

Reply:
This has led to some downstream complications in our internal code. Namely, we have use-cases when we unlink an item from the main dictionary and free it afterwards. When the actual free happens we can't assume the `dict` itself still exists as now we allocate dict on demand, so the `dict` itself can be freed if user performs flush operation. 

Conceptually it doesn't really make sense to require the `dict` itself to be present when we free the key and value.

Reply:
That's true, but we don't want to structure the code to account for downstream patches. So far, none of the dict functions take a dictType, apart from dictCreate. We are making the API a bit messy by changing this.

We are using two-phase unlink in Valkey, for example in kvstore.c:

```C
dictEntry *
kvstoreDictTwoPhaseUnlinkFind(kvstore *kvs, int didx, const void *key, dictEntry ***plink, int *table_index) {
    dict *d = kvstoreGetDict(kvs, didx);
    if (!d) return NULL;
    return dictTwoPhaseUnlinkFind(kvstoreGetDict(kvs, didx), key, plink, table_index);
}

void kvstoreDictTwoPhaseUnlinkFree(kvstore *kvs, int didx, dictEntry *he, dictEntry **plink, int table_index) {
    dict *d = kvstoreGetDict(kvs, didx);
    dictTwoPhaseUnlinkFree(d, he, plink, table_index);
    cumulativeKeyCountAdd(kvs, didx, -1);
    freeDictIfNeeded(kvs, didx);
}
```

Here we delete dict only after it has been unlinked and also freed.

Can you explain what you do differently in your downstream patches and why? Is it for data tiering or something like that?

Reply:
It is true that currently in OSS there is no use-case when you hold onto an entry that is unlinked from the main dict and later frees it. Internally we have such a use-case for forkless save. It has to do with passing the item to a background thread to process and later free it. 

`dictFreeKey()` and `dictFreeVal()` are no longer API methods as defined in `dict.h` so this change should not make the API messy. 

However, this reminds me that the API `dictFreeUnlinkedEntry()` in `dict.h` is still taking in `dict` as argument. It has no good reason to as the entry is already unlinked from the dict itself. I think maybe I should change it to 

```
void dictFreeUnlinkedEntry(dictType *type, dictEntry *he) {
    if (he == NULL) return;
    dictFreeKey(type, he); 
    dictFreeVal(type, he);
    /* Clear the dictEntry */
    if (!entryIsKey(he)) zfree(decodeMaskedPtr(he));
}
```

Reply:
> `dictFreeKey()` and `dictFreeVal()` are no longer API methods as defined in `dict.h` so this change should not be making the API messy.

True. We make these internal in dict.c now, so it's no disaster to change them to whatever we want. But there's also no reason to change them. Why would we change 1 to 2, and not all the way to 3 or 4?

1. `dictFreeVal(d, de);`
2. `dictFreeVal(d->type, de);`
3. `dictFreeVal(d->type->valDestructor, de);`
4. `if (d->type->valDestructor) d->type->valDestructor(dictGetVal(de))`;

Your downstream patches can actually use 4 directly.

`dictCompareKeys()` still is in the public API. It's used from memory prefetch only, right? I wouldn't mind we make `dictCompareKeys()` internal in dict.c and the memory prefetch can just use method 4 above, i.e. just call the `d->type->keyCompare` function directly. Or don't change it at all. :) Not changing is safe if you want to avoid a debate.

Reply:
For now I will leave `dictCompareKeys()` in `dict.h` to avoid a debate. 

I think it is a good idea to just pass in the `keyDestructor` into `dictFreeKey()`, and `valDestructor` into `dictFreeVal()`. Will explore that along with `void dictFreeUnlinkedEntry(dictType *type, dictEntry *he)`

Reply:
To avoid the debate about dictCompareKeys, you'd need to change it back to what it was though.

Reply:
I feel we should not be invoking `dict->type->valDestructor` directly in our internal code to free an unlinked entry. Rather it should go through the API `void dictFreeUnlinkedEntry(dictType *type, dictEntry *he)`. 

To avoid code duplications, I think I will still keep `dictCompareKeys()` to be in `dict.h`. However, I am thinking of changing it from a macro into a `static inline int dictCompareKeys(dictType *type, const void *key1, const void *key2)`.